### ISSUE TITLE
useAsyncTaskFetch: pending is "true" from the start

### ISSUE THREAD

**Issue by bugleev** [dai-shi/react-hooks-async#22]
Hi, noticed some weird behaviour when firing useAsyncTaskFetch by using the **.start()** method.

**Pending** status is set to true when fetch is not started, and also **.started** is not set to false after fetch is complete.

Here's demo:
https://codesandbox.io/s/react-hooks-async-demo-ebe5n
You can see loading state from the start, and button is disabled after fetch is done.

Also, maybe it's codesandbox related, but after installing "react-hooks-async" dependency, there was an error 

> Could not find dependency: 'core-js' relative to '/node_modules/react-hooks-async/dist/index.js'

so i had to add it manually.

**ADDED 30.08.19**
Found  one more thing. 
If you remove disabled attribute of the button in that codesanbox demo, and fire another fetch while previous is running (better do it in slow3g mode), the request is canceled and another one is started, that's ok, but the task has **pending** status **false** . 
I looked into **use-async-task.j**s and it looks like after firing next request, 
`forceUpdateRef.current(func);` of the previous task object runs **after** `forceUpdateRef.current(func);` of the next task. So we get pending state of the finished task.
As a workaround i added this to the task object:
```
safeStart: () => {
      task.abort();
      setTimeout(() => {
        task.start();
      }, 0);
```
and exporting it as a start method 
```
  return useMemo(
    () => ({
      start: task.safeStart,
```
and it worked ok, JFYI

**Comment by dai-shi** [comment]
Thanks so much for reporting this. I’d admit I wasn’t considering deeply enough when I did #18. Give me a couple of days. I’m afraid this is going to be a breaking change technically. I wonder if we can do minor version up.

**Comment by dai-shi** [comment]
As for core-js, it's expected that users need to install it by themselves depending on their environments. I'm not sure if there's a better way.

**Comment by dai-shi** [comment]
> ADDED 30.08.19
Found one more thing.

This is obviously a bug. I will work on it later.

Meanwhile, what do you think is the expected behavior of `started` and `pending` flags?
Originally, `started` exists to check if we can call `start()`, but after #18 we can call `start()` anytime (and the previous task will be aborted as you observed.)

**Comment by dai-shi** [comment]
What we could do currently is `asyncTask.started && asyncTask.pending`.
https://codesandbox.io/s/react-hooks-async-demo-e8wko

That seems a bit weird.

I'd propose removing `.started` completely and setting `pending = false` at start.
Do you have any concerns with it?

**Comment by dai-shi** [comment]
b2da62a should fix the bug of overwriting by the previous task.
@bugleev Could you try it to see if it fixes your use case?

**Comment by bugleev** [comment]
> I'd propose removing .started completely and setting pending = false at start.
Do you have any concerns with it?

Sure, 100% should be removed then. No idea why i used it in the first place, was just figuring stuff out.
> Could you try it to see if it fixes your use case?

Will do

**Comment by bugleev** [comment]
https://github.com/dai-shi/react-hooks-async/commit/b2da62ae59731c9dc02942c99c724898cdd0ea99 works just fine! `Pending` should be false from the start, as you mentioned.

**Comment by dai-shi** [comment]
Thanks for the confirmation.

As I started thinking about setting pending=false, I have one concern.
In TS types, we define:
```typescript
export type AsyncTask<Result, Args extends unknown[] = unknown[]> = {
  start: (...args: Args) => void;
  abort: () => void;
  started: boolean;
} & (
  | {
    pending: true;
    error: null;
    result: null;
  }
  | {
    pending: false;
    error: Error;
    result: null;
  }
  | {
    pending: false;
    error: null;
    result: Result;
  }
);
```

This ensures that if I check `pending` is false and `error` is null, `result` is always available.
This behavior eases coding in TypeScript.

If we want to keep this behavior, we keep `.started` flag and clear it when it's fulfilled.
So,
- initial: started=false, pending=true
- first start: started=true, pending=true
- first end: started=false, pending=false
- second start: started=true, pending=true
- second end: started=false, pending=false

Well, this might not be intuitive. Hmm.

**Comment by bugleev** [comment]
I'm sorry, I don't have much experience in typescript. But two things bother me here: 
1. Designing API from the perspective of typescript-only usage;
2. It feels odd to me, that task is pending before it's even started;

**Comment by dai-shi** [comment]
Those are totally fair comments.

I'll give another example. Let's assume a simple usage in which we don't manually call `.start()`.
```javascript
import { useFetch } from 'react-hooks-async';
const Component = () => {
  const { pending, result } = useFetch('https://yesno.wtf/api');
  if (pending) return 'Loading...';
  if (!result) return 'Not started';
  return result.answer;
};
```

I didn't run this, but if we run this, we would see `Not started` at the first time, and then `Loading...`. This is because we invoke `.start()` in useEffect.

To solve this with pending = false at the beginning, we would need to do this:
```javascript
import { useFetch } from 'react-hooks-async';
const Component = () => {
  const { pending, result } = useFetch('https://yesno.wtf/api');
  if (pending || !result) return 'Loading...';
  return result.answer;
};
```
This isn't intuitive either. (furthermore, result could be `null` or `0` in a certain use case.)

@bugleev Do you get the point?

**Comment by dai-shi** [comment]
f092d4a this commit resets `.started` after fulfilled.
So, in your use case, you can only use `.started`.

Having two flags is obviously confusing, but I don't yet find another solution.

**Comment by bugleev** [comment]
Ok, i get the point, although personally i don't see much value in 'Not started' state. We either render spinner or we render data component, where we check if there's data or not.
I usually do something like this:

```javascript
const Component = () => {
  const { pending, result } = useFetch("https://yesno.wtf/api");
  const renderSpinner = "loading";
  const renderData = result ? result.answer : "no data";
  return pending ? renderSpinner : renderData;
};
```
But anyway, as far as .start() works now with .started flag, that's good enough for me)

**Comment by dai-shi** [comment]
Thanks again for your feedback.

Unfortunately, your modified example still shows `no data` at first.

OK, I will keep two flags for now. Once we have a better idea, we will re-consider.
Thanks for raising this up.

**Comment by dai-shi** [comment]
Published: https://www.npmjs.com/package/react-hooks-async/v/3.4.0

I will close this issue. But we still look for better solutions. Any new ideas/suggestions are welcome (it would probably be better to file a new issue).

**Comment by bugleev** [comment]
Thank you for your work! Tools you are creating are so good!
